### ISSUE TITLE
Add support for Windows

### ISSUE THREAD

**Issue by nlhkh** [jodersky/sbt-jni#20]
The command `uname -sm` only works for Unix-based OSes. Currently I have to manually set
```scala
nativePlatform := "win32-x86_64"
```

**Comment by jodersky** [comment]
Windows is currently not officially supported, however it shouldn't be too complicated to add. Here is the rough outline:
- find a way to distinguish the windows architectures (I'm not sure if there are mutliple), similar to what uname can provide on unix
- use that mechanism in [JniNative](https://github.com/jodersky/sbt-jni/blob/master/plugin/src/main/scala/ch/jodersky/sbt/jni/plugins/JniNative.scala#L40) and [JniLoad](https://github.com/jodersky/sbt-jni/blob/master/macros/src/main/scala/ch/jodersky/jni/annotations.scala#L51) to bundle and extract the corresponding native libraries

**Comment by eaplatanios** [comment]
What about `System.getProperty("os.arch")` and `System.getProperty("os.name")`. Could we change the code to use these instead of `uname`?

**Comment by jodersky** [comment]
That's a good idea and was also what was used in a very early version of this project. Unfortunately, the `os.arch` property is too vague. It is able to distinguish between certain architectures only, such as x86 and ARM for instance. The latter however has several, incompatible, architectures which are indistinguishable with that mechanism. For example armv6 (used in the original raspberry pi) and armv7l (used in subsequent models) are incompatible, yet they both show up as simply "arm" from `os.arch`.

**Comment by eaplatanios** [comment]
I see. How about we fall back to these two properties when `uname` is not available. Do you think that's acceptable behavior until we find a better solution?

**Comment by jodersky** [comment]
I think we could check the `os.name` property and in case it is windows handle it specifically. Otherwise we should use the standard uname method.

**Comment by andraxin** [comment]
My platform is neither Windows nor ARM, but I'm still inconvenienced by the need to spawn an extra process to run `uname` **during JNI loading**, since that's not allowed causing my application to crash.
My fix was also to use `sys.props("os.name")` and `sys.props("os.arch")` to avoid forking a process.

> That's a good idea and was also what was used in a very early version of this project. Unfortunately, the `os.arch` property is too vague. It is able to distinguish between certain architectures only, such as x86 and ARM for instance. The latter however has several, incompatible, architectures which are indistinguishable with that mechanism. For example armv6 (used in the original raspberry pi) and armv7l (used in subsequent models) are incompatible, yet they both show up as simply "arm" from `os.arch`.

I may be jumping to conclusions to fast, but would it be correct to assume that `os.arch` == "arm' is the only problem case?  Shouldn't this be handled via a trouble report back to Oracle to fix this in Java?  That coarse-grain reporting of the underlying architecture is obviously(?) not very helpful and likely to cause pain for others as well...

> I think we could check the `os.name` property and in case it is windows handle it specifically. Otherwise we should use the standard uname method.

I don't know, but dragging in `uname` is only needed on ARM, then -perhaps- special-casing Windows is not the optimal approach.  In my use case, I'd much rather opt for only falling back to `uname` when `os.arch == "arm"`.

What do you think?

**Comment by jodersky** [comment]
I doubt that anything can be done about changing the property of `os.arch` upstream, as that could break user code. However, I think your proposal of using system properties by default and calling uname only on an arm platform is worth a shot!

**Comment by andraxin** [comment]
Great!  Will you fix that or would you prefer a pull request?

**Comment by andraxin** [comment]
[PR](https://github.com/jodersky/sbt-jni/pull/32) created.

**Comment by shado23** [comment]
It's a bit unfortunate that the PR submitted by andraxin was rejected, because assuming the presence of a uname command is not necessarily justified even under Linux, such as when running in a minimal Docker container.

The only thing you can really rely on is that Java is installed. For edge-cases where os.arch is not enough, why not have the user set -Dos.arch=whatever or have some other way to override it.
As far as I can tell neither JNA or JNR use this uname cludge.

**Comment by pomadchin** [comment]
Hey @shado23 I think we can add an option to overrride the target arch via the java option passed, would you like to help with that?

**Comment by chitralverma** [comment]
@pomadchin and @jodersky proposing a alternate solution if it has not been considered already.

Regarding the platform identification,
- currently we are relying on `uname` which doesn't work with windows and has to be manually overridden in the `build.sbt` as mentioned in the OP.
- There were other solutions like relying on `os.arch` but you are correct, it is natively deciphered in java and can lead to non deterministic results.
- My suggestion is to not do either. Actually we only need a value for `nativePlatform` to create directories so why not instead rely on the target triple directly from the 2 supported build tools instead of getting the value from `uname` and then parsing it.
- For rust you can simply set the nativePlatform to `rustc -vV | sed -n 's|host: ||p'` and for cmake we can get it from various places like `make -v`, `gcc -dumpmachine` commands.
- Basically the inference of the `nativePlatform` can be as per the build tool selected and not something that we should construct.

Some other smaller things I identified that block windows support (unrelated to where the `nativePlatform` comes from),
- the path construction is a problem. String interpolation or appending has been done which works on *nix OS but not on windows.
- This can be easily fixed with `java.nio.....Paths`

**Comment by pomadchin** [comment]
hey @chitralverma took some time to respond; I think it is a great idea and worth trying :+1:

**Comment by chitralverma** [comment]
Ref hacky project on how this will look like,
https://github.com/chitralverma/test-native-proj

**Comment by pomadchin** [comment]
@chitralverma hmmm, I think it's worth trying starting a PR against this repo just to start iterating on this idea.

**Comment by Martomate** [comment]
Hello!

I just tried to build using Cargo on Windows (in Git Bash, so MinGW) and the build completed successfully but then this plugin could not find the DLL file. This seems to be caused by [this line](https://github.com/sbt/sbt-jni/blob/39d4d724e2296b760ba3301712b7c3bf6d2c5035/plugin/src/main/scala/com/github/sbt/jni/build/Cargo.scala#L45):
```scala
val products: List[File] =
        (targetDirectory / subdir * ("*.so" | "*.dylib")).get().filter(_.isFile).toList
```

Could we include `"*.dll"` in that list? Are there other places where similar changes would be needed?

I can confirm that there was only one DLL file generated in that directory, so it should work.

**Comment by pomadchin** [comment]
👋 hey @Martomate, yes; probably makes sense to include it for all the build tools :+1:

**Comment by Martomate** [comment]
Hello everyone! I just made a [PR](https://github.com/sbt/sbt-jni/pull/234) to remove the need for the `uname` command. It uses the `os.name` and `os.arch` properties, which should be fine most of the time.

Regarding the issue where both `armv6` and `armv7l` are called "arm": could we find a way to support that as an override instead of using `uname`? Maybe it's enough to just set the `nativePlatform` property? Or maybe make the `uname` solution opt-in?

**Comment by pomadchin** [comment]
@Martomate I think uname should be the default if it's available, in other cases relying on os.name / os.arch seems to be good enough. 

Another interesting alternative is https://github.com/sbt/sbt-jni/issues/20#issuecomment-1413763299
